packages/edge-login-ui-rn/src/components/themed/FourDigitInput.js

```diff
@@ -0,0 +1,143 @@
+// @flow
+
+import * as React from 'react'
+import {
+  ActivityIndicator,
+  TextInput,
+  TouchableWithoutFeedback,
+  View
+} from 'react-native'
+import { cacheStyles } from 'react-native-patina'
+
+import { validatePin } from '../../actions/CreateAccountActions'
+import type { Dispatch, RootState } from '../../types/ReduxTypes'
+import { useRef } from '../../util/hooks'
+import { fixSides, mapSides, sidesToMargin } from '../../util/sides'
+import { connect } from '../services/ReduxStore'
+import {
+  type Theme,
+  type ThemeProps,
+  withTheme
+} from '../services/ThemeContext'
+
+export const MAX_PIN_LENGTH = 4
+
+type OwnProps = {
+  marginRem?: number[] | number
+}
+
+type StateProps = {
+  pin: string,
+  wait: number
+}
+
+type DispatchProps = {
+  onChangeText(pin: string): void
+}
+
+type Props = OwnProps & StateProps & DispatchProps & ThemeProps
+
+const FourDigitInputComponent = ({
+  pin = '',
+  wait,
+  marginRem,
+  theme,
+  onChangeText
+}: Props) => {
+  const inputRef = useRef<TextInput | null>(null)
+  const spacings = sidesToMargin(mapSides(fixSides(marginRem, 0.5), theme.rem))
+  const styles = getStyles(theme)
+
+  const handleRefocus = () => {
+    if (inputRef.current !== null) inputRef.current.focus()
+  }
+
+  const handleUpdate = (pin: string) => {
+    // Change pin only when input are numbers
+    if (/^\d+$/.test(pin) || pin.length === 0) {
+      onChangeText(pin)
+    }
+  }
+
+  return (
+    <TouchableWithoutFeedback onPress={handleRefocus}>
+      <View style={[styles.container, spacings]}>
+        <View style={styles.interactiveContainer}>
+          {wait > 0 ? (
+            <ActivityIndicator color={theme.iconTappable} size="large" />
+          ) : (
+            <View style={styles.dotContainer}>
```

Comment:
You should try and avoid writing code like this.
When you have a repeating pattern, you should generalize it like here: 

https://github.com/EdgeApp/edge-react-gui/blob/develop/src/components/themed/PinDots.js

And in this case you should also separate the "PinDots" component like the one in the link.